lib/services/join.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package services
+
+import (
+	"context"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+)
+
+// JoinService abstracts the proto.JoinService interface so that it can by
+// implemented by both the auth client and the auth server.
+type JoinService interface {
+
+	// RegisterUsingIAMMethod registers the caller using the IAM join method and
+	// returns signed certs to join the cluster.
+	//
+	// The caller must provide a ChallengeResponseFunc which returns a
+	// *types.RegisterUsingTokenRequest with a signed sts:GetCallerIdentity
+	// request including the challenge as a signed header.
+	RegisterUsingIAMMethod(ctx context.Context, challengeResponse types.RegisterChallengeResponseFunc) (*proto.Certs, error)
+}
+
+// JoinServiceGRPCServer implements proto.JoinServiceServer and is designed
+// to run on both the Teleport Proxy and Auth servers.
+type JoinServiceGRPCServer struct {
```

Comment:
Does this have to be under `lib/services`? Could we move it to a more specific package?

Our catch-all packages, like lib/auth and lib/services, are already too big.

Reply:
Moved to new package `lib/joinserver`